src/service/position-observer-impl.js

```diff
@@ -0,0 +1,410 @@
+  layoutRectEquals,
+  layoutRectsOverlap,
+  layoutRectFromDomRect,
+  layoutRectLtwh,
+} from '../layout-rect';
+import {isExperimentOn} from '../../src/experiments';
+
+/** @const @private */
+const TAG = 'POSITION_OBSERVER';
+
+/**
+ * The positionObserver returned position value which includes the position rect
+ * relative to viewport. And viewport rect which always has top 0, left 0, and
+ * viewport width and height.
+ * @typedef {{
+ *  positionRect: ?../layout-rect.LayoutRectDef,
+ *  viewportRect: !../layout-rect.LayoutRectDef,
+ * }}
+ */
+export let PositionInViewportEntryDef;
+
+/** @enum {number} */
+export const PositionObserverFidelity = {
+  HIGH: 1,
+  LOW: 0,
+};
+
+/** @const @private */
+const LOW_FEDELITY_FRAME_COUNT = 2;
+
+class AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+    /** @const {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {!Array<!Object>} */
+    this.entries_ = [];
+
+    /** @private {!./vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @private {!./viewport-impl.Viewport} */
+    this.viewport_ = viewportForDoc(ampdoc);
+
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   * @param {function(PositionInViewportEntryDef)} handler
+   */
+  observe(element, fidelity, handler) {
+    // TODO(@zhouyx, #9208) make entry into a class
+    const entry = {
+      element,
+      handler,
+      fidelity,
+      position: null,
+      turn: (fidelity == PositionObserverFidelity.LOW) ?
+          Math.floor(Math.random() * LOW_FEDELITY_FRAME_COUNT) : 0,
+      trigger: function(position) {
+        const prePos = entry.position;
+        if (prePos
+            && layoutRectEquals(prePos.positionRect, position.positionRect)
+            && layoutRectEquals(prePos.viewportRect, position.viewportRect)) {
+          // position doesn't change, do nothing.
+          return;
+        }
+        if (layoutRectsOverlap(position.positionRect, position.viewportRect)) {
+          entry.position = position;
+          // Only call handler if entry element overlap with viewport.
+          try {
+            entry.handler(position);
+          } catch (err) {
+            // TODO(@zhouyx, #9208) Throw error. QQQ dev or user?
+          }
+        } else if (entry.position) {
+          // Need to notify that element gets outside viewport
+          // NOTE: This is required for inabox position observer.
+          entry.position = null;
+          position.positionRect = null;
+          entry.handler(position);
+        }
+      },
+    };
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback();
+    }
+
+    this.vsync_.measure(() => {
+      this.updateEntryPosition_(entry);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
```

Comment:
Who would call `changeFidelity`? Is this a real API or a `@protected` API? I'd assume different clients would just call `observe` for the same element, but with different fidelity?